MEP-0001: Future-based shuffle on Ray (#2)
mars-project/meps: mep-0001-ray-future-based-shuffle.md

# Motivation
Shuffle is the core operation for large-scale data processing. The data is partitioned according to the partition key in each record through the ALL-TO-ALL connection, and sent from all mappers to all reducers. Due to the nature of its ALL-TO_ALL connections, shuffle often becomes the bottleneck of the performance, stability and scalability of the entire system. This MEP introduces a pull-based shuffle implementation using ray future which serves following goals:
- Improve Mars-on-Ray shuffle scalability by reducing shuffle meta
- Improve scheduling performance by using ray c++ scheduler
- Support shuffle failover by lineage recontruction
- Setup the foundation for push-based shuffle using ray future

## Design and Architecture
When the Mars subtask has upstream and downstream dependencies, the downstream subtask's dependence on the output of the upstream subtask is determined by the data key. The input data required for subtask execution is a context dict object from data key to data. 

For non-shuffle operators, the data key is the chunk key; but for shuffle operators, the data key is a tuple of (chunk key, reducer index). 

> Therefore, in order to build a Shuffle DAG based on Ray future, either all the data keys output by each mapper subtask are known in advance, or the data keys are removed.

Reviewer: ```suggestion
Therefore, in order to build a Shuffle DAG based on Ray future, either all the data keys output by each mapper subtask are known in advance, or the data keys can be calculated from the subtask (chunk graph).
```

### Building shuffle DAG by reducer index

> Acutally, Mars' ChunkGraph has information about the upstream and downstream chunk dependencies. After converting to Ray DAG, this part of the information still exists. The reason for relying on data keys is that Mars replaced `ShuffleProxy` with `FetchShuffle`. The implementation of `Fetch/FetchShuffle` is an abstraction of P2P, which is actually inconsistent with the semantics of Ray DAG. Therefore, when using Ray Backend, a new set of Shuffle operators can be introduced to remove FetchShuffle and rely on Ray to automatically resolve subtask input arguments.

Reviewer: ```suggestion
Acutally, Mars' ChunkGraph has information about the upstream and downstream chunk dependencies. After converting to Ray DAG, this part of the information still exists. The reason for relying on data keys is that Mars replaced `ShuffleProxy` with `FetchShuffle`. The implementation of `Fetch` or `FetchShuffle` is an abstraction of P2P, which is actually inconsistent with the semantics of Ray DAG. Therefore, when using Ray Backend, a new set of Shuffle operators can be introduced to replace the `FetchShuffle` and rely on Ray to automatically resolve subtask dependencies.
```

#### Operand Changes
- Add `n_reducers` and `reducer_ordinal` to all `MapReduceOperand` operands
- Introduce a new group of `FetchShuffle` operands which only records `n_mappers` and `n_reducers` to reduce meta overhead:
```python
class FetchShuffleType(Enum):
    FETCH_SHUFFLE_BY_INDEX = 0
    FETCH_SHUFFLE_BY_DATA_KEY = 1

class FetchShuffleByIndex(Operand):
    _op_type_ = opcodes.FETCH_SHUFFLE_BY_INDEX
    # number of all upstream mappers
    n_mappers = Int32Field("n_mappers")
    # number of all downstream reducers
    n_reducers = Int32Field("n_reducers")

class FetchShuffle(Operand):
    _op_type_ = opcodes.FETCH_SHUFFLE_BY_DATA_KEY
    source_keys = ListField("source_keys", FieldTypes.string)
    source_idxes = ListField("source_idxes", FieldTypes.tuple(FieldTypes.uint64))
    source_mappers = ListField("source_mappers", FieldTypes.uint16)
```
- Make shuffle operands mapper produce output with deterministic number and order.
    - `TensorReshape` and `TensorBinCount` mapper needs change. 
- _bagging.py uses `source_idxes`, which alsoe needs to be refactored.

#### Graph Changes

> Add an option to replace `ShuffleProxy` to `FetchShuffleByIndex` when building subtask graph

Reviewer: ```suggestion
Add an option to replace `ShuffleProxy` to `FetchShuffleByIndex` when building subtask graph.
```

#### Using a ShuffleManager to manage Ray shuffle execution

> `ShuffleManager` is defined in `mars/services/task/execution/ray/shuffle.py` and only used by ray executor. When `TaskProcessorActor` invokes `RayTaskExecutor.execute_subtask_graph` to execute a subtask graph, `RayTaskExecutor` will create `ShuffleManager` to manage shuffle execution.

Reviewer: ```suggestion
`ShuffleManager` is defined in `mars/services/task/execution/ray/shuffle.py` and only used by ray executor. When `TaskProcessorActor` invokes `RayTaskExecutor.execute_subtask_graph` to execute a subtask graph, the `RayTaskExecutor` will create a `ShuffleManager` to manage shuffle execution.
```

`ShuffleManager` will do following things:
- Build mapper index which is a dict from mapper subtask to (`shuffle_index`, `mapper_index`). A subtask graoh may have multiple groups of shuffle, `shuffle_index` indicates a mapper subtask belongs to which shuffle. `mapper_index` indicates is oridinal in all mapper subtasks of current shuffle.

> - Build reducer index which is a dict from reducer subtask to (`shuffle_index`, `reducer_ordinal`). A subtask graoh may have multiple groups of shuffle, `shuffle_index` indicates a reducer subtask belongs to which shuffle. `reducer_ordinal` indicates is oridinal in all reducer subtasks of current shuffle. If some reducers missing, `ShuffleManager` will filling None for those reducers index.

Reviewer: ```suggestion
- Build reducer index which is a dict from reducer subtask to (`shuffle_index`, `reducer_ordinal`). A subtask graoh may have multiple groups of shuffle, `shuffle_index` indicates a reducer subtask belongs to which shuffle. `reducer_ordinal` indicates is oridinal in all reducer subtasks of current shuffle. If some reducers are missing, `ShuffleManager` will fill `None` for those reducers index.
```

> - Recording mapper output object refs for all mapper subtasks, which will be used by later reducers get all mapper inputs based on reducer oridnal.

Reviewer: ```suggestion
- Recording mapper output object refs for all mapper subtasks, which will be used by later reducers to get all the mapper inputs based on reducer oridnal.
```

Author: `all mapper inputs` is more idiomatic

- Return `n_reducers` for a shuffle when passing a mapper/reducer subtask.
- In the future, it will manage push-based shuffle scheduling and execution.
 
The code skeleton is as follows:
 
```python
class ShuffleManager:
    """Manage shuffle execution for ray by resolve dependencies between mappers outputs and reducers inputs based on
    mapper and reducer index.
    """

    def __init__(self, subtask_graph):
        self.subtask_graph = subtask_graph
        # Build mapper index
        # subtask: (shuffle_index, mapper_index)
        # Build reducer index
        # subtask: (shuffle_index, reducer_ordinal)

    def has_shuffle(self) -> bool:
        """
        Returns
        -------
        bool
            Whether current subtask graph has shuffle to execute
        """
        return self.num_shuffles > 0

    def add_mapper_output_refs(self, subtask, output_object_refs: List["ray.ObjectRef"]):
        """
        Record mapper output ObjectRefs which will be used by reducers later.

        Parameters
        ----------
        subtask
        output_object_refs : List["ray.ObjectRef"]
            Mapper output ObjectRefs.
        """
        pass

>     def get_reducer_input_refs(self, subtask: Subtask) -> List["ray.ObjectRef"]:

Reviewer: Maybe we can move the `get_reducer_input_refs` and `get_n_reducers` to the `SubtaskGraph` in the future, then Ray backend can share the same logic with Mars backend.

Author: Actually, those refs are Ray ObjectRefs, and in future there will be more ray-specific Ray Object management  code when push-based shuffle is implemented for ray. I think it's not feasible to share this logic with Mars backend

Reviewer: I know the refs are Ray object refs. For Mars, they are the map data keys. Mars can use this logic to calculate the map data keys.

Author: When push-based shuffle are implemented, parts of those refs are mapper ouputs, parts of it are merge outputs. I think those logic shouldn't be put into SubtaskGraph

# Motivation

> Shuffle is the core operation for large-scale data processing. The data is partitioned according to the partition key in each record through the ALL-TO-ALL connection, and sent from all mappers to all reducers. Due to the nature of its ALL-TO_ALL connections, shuffle often becomes the bottleneck of the performance, stability and scalability of the entire system. This MEP introduces a pull-based shuffle implementation using Ray future which serves following goals:

Reviewer 2: ```suggestion
Shuffle is the core operation for large-scale data processing. The data is partitioned according to the partition key in each record through the ALL-TO-ALL connection, and sent from all mappers to all reducers. Due to the nature of its ALL-TO-ALL connections, shuffle often becomes the bottleneck of the performance, stability and scalability of the entire system. This MEP introduces a pull-based shuffle implementation using Ray future which serves following goals:
```

 
#### Operand Changes
- Add `n_reducers` and `reducer_ordinal` to all `MapReduceOperand` operands
- Introduce a new group of `FetchShuffle` operands which only records `n_mappers` and `n_reducers` to reduce meta overhead:
```python
class FetchShuffleType(Enum):
   FETCH_SHUFFLE_BY_INDEX = 0
   FETCH_SHUFFLE_BY_DATA_KEY = 1
 
class FetchShuffleByIndex(Operand):
   _op_type_ = opcodes.FETCH_SHUFFLE_BY_INDEX
   # number of all upstream mappers
   n_mappers = Int32Field("n_mappers")
   # number of all downstream reducers
   n_reducers = Int32Field("n_reducers")
 
class FetchShuffle(Operand):
   _op_type_ = opcodes.FETCH_SHUFFLE_BY_DATA_KEY
   source_keys = ListField("source_keys", FieldTypes.string)
   source_idxes = ListField("source_idxes", FieldTypes.tuple(FieldTypes.uint64))

>    source_mappers = ListField("source_mappers", FieldTypes.uint16)

Reviewer 2: Use `FetchShuffleByIndex` to replace `FetchShuffle`? If so, `FetchShuffle` can be removed.

Author: `FetchShuffle` will still be used by mars shuffle

Reviewer 3: Can we just record the n_mappers and n_reducers in `FetchShuffle`? and I think for Mars shuffle we can remove the `source_keys` stuff as well, but for compatibility we can mix them together, I think we don't need a new opcode.

Author: If mars can remove `source_keys ` stuff too, add `n_mappers` and `n_reducers` would be better.
